### ISSUE TITLE
For multi-pass rendering, add "multiple cameras" to Viewport, "depth" to Camera, and "keep" and "keep depth"

### ISSUE THREAD

**Issue by 31** [godotengine/godot-proposals#1428]
This is an implementation proposal for https://github.com/godotengine/godot-proposals/issues/956 and https://github.com/godotengine/godot-proposals/issues/932. I figure it's worth filing separately, so that if there's no way it'll happen this way, we can close this proposal without derailing the more general ones.

**Describe the project you are working on:**

* An FPS game with a gun viewmodel.
  * It's common practice to render this separately from the world, with different FoV, near/far clip planes. Using cull masks on the multiple cameras, this can be done.
  * Described more at https://github.com/godotengine/godot-proposals/issues/956.
* A VR game with an effect that requires multiple render passes.
  * Described more at https://github.com/godotengine/godot-proposals/issues/932. That issue proposes a different implementation, but it can be accomplished just as well using multiple cameras.

**Describe the problem or limitation you are having in your project:**

I can use Viewports to accomplish this, but that **doesn't work at all in VR** (see https://github.com/godotengine/godot-proposals/issues/932), and is a (unknown severity) performance issue.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

These purely additive new properties (to avoid compat problems) work together to make both scenarios possible:

* Viewport
  * `use_multiple_cameras` - Enables multi camera mode. This tells the renderer to use `cameras` rather than the single `current` `camera`.
* Camera
  * `depth` - This determines the order that cameras are rendered in when `use_multiple_cameras` is enabled. No effect otherwise. If two cameras have the same `depth`, behavior is undefined.
    * This property is familiar to those who have used Unity and Unreal, where this feature already exists.
* Environment
  * `keep` - Tells the renderer not to clear the current color when rendering the camera with this environment.
    * We can't simply use `mode` `KEEP`, because that doesn't let us use sky-contributed lighting. This means `keep` is redundant, but to keep these features purely additive, this is what I came up with.
  * `keep_depth` - Tells the renderer not to clear the depth buffer. This is not desired for the FPS viewmodel scenario, but is required for the VR effect I want to accomplish.

I would prefer to have the new Environment properties on Camera. That would let me reuse the same Environment for multiple Cameras, rather than creating a duplicated copy for each Camera just to change the one bool. However, Environment data is already passed deep into the renderer, and it's easier to code this way. If someone more skilled than me takes this over, putting these on Camera would be a lot better. 😄 It's difficult for me to tell when changing function signatures would be a massive breaking change for developers of C++ extensions and whatnot.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

I tried implementing it in 3.2 as a proof of concept, which works:
Godot changes: [31:multicam/3.2](https://github.com/godotengine/godot/compare/3.2...31:multicam/3.2)
Example use: https://github.com/31/example-arvr-viewports/blob/vr-custom/main.tscn

That branch is not done:
* It's only implemented in the GLES3 renderer.
* There's an error that pops up repeatedly that I assume is due to some code depending on there being only one `current` camera or some missing code with how I added multiple cameras. (Although the test scenes work, I assume some Camera-related stuff is broken.)
  ```
  ERROR: SpatialIndexer::_update_camera: Condition "!E" is true.
     At: scene\resources\world.cpp:126
  ```
* You have to be careful to not set `current` on the last camera to render. Don't know why.
* I disabled a VR optimization when `use_multiple_cameras` because it breaks multi-cam `cull mask`s on the right eye.
* There's an arbitrary `VIEWPORT_CAMERAS_MAX = 32` with no indication to the user. (It's potentially unnecessary, I'm just not sure if `Vector<T>` is permitted in server data--could only find arrays.)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
**Is there a reason why this should be core and not an add-on in the asset library?:**

This can't be implemented as a workaround or addon: the renderer is not flexible enough to allow it. I've heard that the 4.0 pipeline should be more flexible, but last I heard there were no concrete plans, and it would be nice to get it before then.

**Comment by 31** [comment]
Chatted with Calinou on IRC and this doesn't seem like it would fit in `3.2`.

Ideally we can use scriptable render pipelines instead (https://github.com/godotengine/godot-proposals/issues/644?) but I don't know whether that's going to be in 4.0... so it might still be useful to get this into `master`.

**Comment by aaronfranke** [comment]
See also: https://github.com/godotengine/godot/pull/8215

**Comment by atirut-w** [comment]
Any chance this would be added to 3.x?

**Comment by aaronfranke** [comment]
@atirut-w No. Rendering features are developed on master (what will become Godot 4.0), and it has a very different codebase from Godot 3.x. Adding this to Godot 3.x would involve making the feature twice (and at the moment it hasn't even been made once yet).
